File: packages/auth/ios/RNFBAuth/RNFBAuthListenerRegistry.m
/**
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "RNFBAuthListenerRegistry.h"

#if __has_include("RNFBHandleMap.h")
#import "RNFBHandleMap.h"
#else
#import "RNFBApp/RNFBHandleMap.h"
#endif

@protocol RNFBAuthRemovable <NSObject>
- (void)remove;
@end

@interface RNFBAuthListenerRegistry ()
@property(nonatomic, strong) RNFBHandleMap *map;
@end

@implementation RNFBAuthListenerRegistry

- (instancetype)init {
  self = [super init];
  if (self) {
    _map = [[RNFBHandleMap alloc] init];
  }
  return self;
}

- (void)rnfb_removeHandle:(id)handle {
  if (handle && [handle respondsToSelector:@selector(remove)]) {
    [(id<RNFBAuthRemovable>)handle remove];
  }
}

- (BOOL)put:(id)key value:(id)value error:(NSError **)error {
  return [self.map put:key value:value error:error];
}

- (BOOL)putOrDiscard:(id)key value:(id)value {
  NSError *error = nil;
  if ([self.map put:key value:value error:&error]) {
    return YES;
  }
  [self rnfb_removeHandle:value];
  return NO;
}

- (id)get:(id)key {
  return [self.map get:key];
}

- (id)take:(id)key {
  return [self.map take:key];
}

- (void)takeAndRemove:(id)key {
  id handle = [self.map take:key];
  [self rnfb_removeHandle:handle];
}

- (void)removeAll {
  NSArray *handlers = [self.map takeAll];
  for (id handler in handlers) {
    [self rnfb_removeHandle:handler];
  }
}

@end

File: packages/auth/android/src/main/java/io/invertase/firebase/auth/RNFBAuthListenerRegistry.java
package io.invertase.firebase.auth;

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import io.invertase.firebase.common.RNFBHandleCollisionException;
import io.invertase.firebase.common.RNFBHandleMap;
import java.util.List;

/**
 * Auth-state / id-token listener map. Unique {@link #put} / {@link #putOrDiscard}; callers {@link
 * #take} or {@link #takeAllAndRemove} then {@code remove()} outside the HandleMap lock.
 * Skip-if-registered is {@code get(appName) != null}.
 */
final class RNFBAuthListenerRegistry {
  private final RNFBHandleMap<String, AuthListenerHandle> map = new RNFBHandleMap<>();

  void put(String appName, AuthListenerHandle handle) throws RNFBHandleCollisionException {
    map.put(appName, handle);
  }

  /**
   * Unique put. On collision, {@code remove()} the incoming handle and leave the existing mapping.
   *
   * @return true if stored
   */
  boolean putOrDiscard(String appName, AuthListenerHandle handle) {
    try {
      map.put(appName, handle);
      return true;
    } catch (RNFBHandleCollisionException collision) {
      if (handle != null) {
        handle.remove();
      }
      return false;
    }
  }

  AuthListenerHandle get(String appName) {
    return map.get(appName);
  }

  AuthListenerHandle take(String appName) {
    return map.take(appName);
  }

  void takeAndRemove(String appName) {
    AuthListenerHandle handle = map.take(appName);
    if (handle != null) {
      handle.remove();
    }
  }

  void takeAllAndRemove() {
    List<AuthListenerHandle> remaining = map.takeAll();
    for (AuthListenerHandle handle : remaining) {
      if (handle != null) {
        handle.remove();
      }
    }
  }
}
